### PR TITLE
Adds UnsafeArrayMap and deprecated BaggageField.getAll for getAllValues

### PR DESCRIPTION
**Pull request by adriancole**
In analyzing the usage of the (recently added BaggageField.getAll), I
noticed in every case the keys are coersed to strings and all non-null
values are read.

This provides a more efficient alternative, via `UnsafeArrayMap`, pulled
in from #1183

### CODE REVIEW

File: brave/src/main/java/brave/internal/baggage/UnsafeArrayMap.java [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+    @Override public boolean removeAll(Collection<?> c) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override public void clear() {
+      throw new UnsupportedOperationException();
+    }
+  }
+
+  static final class Entry<K, V> implements Map.Entry<K, V> {
```
Comment (anuraaga) [inline brave/src/main/java/brave/internal/baggage/UnsafeArrayMap.java:329]: Can we use `SimpleImmutableEntry` instead of this class?

Comment (adriancole) [reply brave/src/main/java/brave/internal/baggage/UnsafeArrayMap.java:329]: sure can. I didn't know that was public! also it is the same as are lowest java level. score!

Comment (adriancole) [reply brave/src/main/java/brave/internal/baggage/UnsafeArrayMap.java]: 80 less lines thanks to you!

**Comment by adriancole** [comment]
you're a champ @anuraaga !

**Pull request merged**